oak/server/enclave_server.h

```diff
@@ -139,61 +141,45 @@ class EnclaveServer final : public asylo::TrustedApplication {
-    void *out_tag;
-    bool ok = false;
-    completion_queue_->Next(&out_tag, &ok);
-    // TODO: Return the value of ok so that we can break the completion queue loop.
-    if (!ok) {
-      LOG(INFO) << "Could not process gRPC event from completion queue";
-    }
+  void RequestNextCall() {
+    // The stream will delete itself when it finishes.
+    auto stream = new grpc_server::GrpcStream(node_);
+    generic_service_.RequestCall(stream->server_context(), stream->server_reader_writer(), completion_queue_.get(), completion_queue_.get(),
+                                 new grpc_server::GrpcEvent(grpc_server::GrpcEvent::NEW_STREAM, stream));
   }
 
   // Consumes gRPC events from the completion queue in an infinite loop.
   void CompletionQueueLoop() {
     LOG(INFO) << "Starting gRPC completion queue loop";
-    int i = 0;
+    RequestNextCall();
     while (true) {
-      ::grpc::GenericServerContext context;
-      ::grpc::GenericServerAsyncReaderWriter stream(&context);
-
-      // We request a new event corresponding to a generic call. This will create an entry in the
-      // completion queue when a new call is available.
-      generic_service_.RequestCall(&context, &stream, completion_queue_.get(),
-                                   completion_queue_.get(), tag(i++));
-
-      // Wait for a generic call to arrive.
-      ProcessNextEvent();
-
-      // Read request data.
-      ::grpc::ByteBuffer request;
-      stream.Read(&request, tag(i++));
-
-      // Process the event related to the read we just requested.
-      ProcessNextEvent();
-
-      ::grpc::ByteBuffer response;
-
-      // Invoke the actual gRPC handler on the Oak Node.
-      ::grpc::Status status = node_->HandleGrpcCall(&context, &request, &response);
-      if (!status.ok()) {
-        LOG(INFO) << "Failed: " << status.error_message();
+      grpc_server::GrpcEvent *tag = nullptr;
+      bool ok = false;
+      if (!completion_queue_->Next(reinterpret_cast<void **>(&tag), &ok)) {
+        return;
       }
 
-      ::grpc::WriteOptions options;
-
-      // Write response data.
-      stream.WriteAndFinish(response, options, status, tag(i++));
-
-      // Process the event related to the write we just requested.
-      ProcessNextEvent();
+      if (tag != nullptr) {
+        auto stream = tag->stream();
+        LOG(INFO) << "Grpc Event:" << tag->event();
+        switch (tag->event()) {
+          case grpc_server::GrpcEvent::NEW_STREAM: {
+            if (ok) {
+              // Indicate to GRPC library that we can handle another call
+              // immediately after this.
+              RequestNextCall();
+            }
+            stream->OnNewGrpc(ok);
+          } break;
+          case grpc_server::GrpcEvent::REQUEST_READ: {
+            stream->OnRequestRead(ok);
+          } break;
+          case grpc_server::GrpcEvent::RESPONSE_WRITTEN: {
+            stream->OnResponseWritten(ok);
+          } break;
+        }
+        delete tag;
```

Comment:
I wonder whether using a `shared_ptr` would let us avoid this manual deallocation (it would effectively use reference-count for the only reference we care about).

Reply:
The gRPC APIs hold a reference to this as a void*, so I'm not sure it is possible to get the benefits of shared_ptr here because there is nothing to keep track of the reference when it is stored as a void*